lib/model/unreads.dart

```diff
@@ -0,0 +1,383 @@
+  /// `true` for [UpdateMessageFlagsAddEvent.all]).
+  bool oldUnreadsMissing;
+
+  final int selfUserId;
+
+  void handleMessageEvent(MessageEvent event) {
+    final message = event.message;
+    if (message.flags.contains(MessageFlag.read)) {
+      return;
+    }
+
+    switch (message) {
+      case StreamMessage():
+        _addLastInStreamTopic(message.id, message.streamId, message.subject);
+      case DmMessage():
+        final narrow = DmNarrow.ofMessage(message, selfUserId: selfUserId);
+        _addLastInDm(message.id, narrow);
+    }
+    if (
+      message.flags.contains(MessageFlag.mentioned)
+      || message.flags.contains(MessageFlag.wildcardMentioned)
+    ) {
+      mentions.add(message.id);
+    }
+    notifyListeners();
+  }
+
+  void handleUpdateMessageEvent(UpdateMessageEvent event) {
+    final messageId = event.messageId;
+    final bool isMentioned = event.flags.any(
+      (f) => f == MessageFlag.mentioned || f == MessageFlag.wildcardMentioned,
+    );
+
+    // We assume this event can't signal a change in a message's 'read' flag.
+    // TODO can it actually though, when it's about messages being moved into an
+    //   unsubscribed stream?
+    //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/mark-as-read.20events.20with.20message.20moves.3F/near/1639957
+    final bool isRead = event.flags.contains(MessageFlag.read);
+    assert(() {
+      if (!oldUnreadsMissing && !event.messageIds.every((messageId) {
+        final isUnreadLocally = _slowIsPresentInDms(messageId) || _slowIsPresentInStreams(messageId);
+        return isUnreadLocally == !isRead;
+      })) {
+        // If this happens, then either:
+        // - the server and client have been out of sync about a message's
+        //   unread state since before this event, or
+        // - this event was unexpectedly used to announce a change in a
+        //   message's 'read' flag.
+        debugLog('Unreads warning: got surprising UpdateMessageEvent');
+      }
+      return true;
+    }());
+
+    switch ((isRead, isMentioned)) {
+      case (true,  false):
+        mentions.remove(messageId);
+      case (true,  true ):
+        // A mention (even if new with this event) makes no difference
+        // for a message that's already read.
+        // TODO skip notifyListeners, unless there's a message move to handle.
+        break;
+      case (false, false):
+        mentions.remove(messageId);
+      case (false, true ):
+        mentions.add(messageId);
+    }
+
+    // TODO: Handle moved messages.
+
+    notifyListeners();
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event) {
+    mentions.removeAll(event.messageIds);
+    switch (event.messageType) {
+      case MessageType.stream:
+        final streamId = event.streamId!;
+        final topic = event.topic!;
+        if (event.messageIds.length == 1) {
+          _removeInStreamTopic(event.messageIds.single, streamId, topic);
```

Comment:
We can skip this optimization and just have the simplicity of always making a `Set` and passing to the remove-all method. That's because this is something we're doing only once for this event, not e.g. once per stream and topic in the data structure.